masonry/src/widgets/canvas.rs

```diff
@@ -0,0 +1,165 @@
+use tracing::{trace_span, Span};
+use vello::kurbo::Size;
+use vello::Scene;
+
+use crate::core::{
+    AccessCtx, AccessEvent, BoxConstraints, EventCtx, LayoutCtx, PaintCtx, PointerEvent, QueryCtx,
+    RegisterCtx, TextEvent, Update, UpdateCtx, Widget, WidgetId, WidgetMut,
+};
+
+/// A widget allowing custom drawing.
+pub struct Canvas {
+    draw: Arc<dyn Fn(&mut Scene, Size) + Send + Sync + 'static>,
+    alt_text: Option<String>,
+}
+
+// --- MARK: BUILDERS ---
+impl Canvas {
+    /// Create a new canvas with the given draw function
+    pub fn new(draw: impl Fn(&mut Scene, Size) + Send + Sync + 'static) -> Self {
+        Self::from_arc(Arc::new(draw))
+    }
+
+    /// Create a new canvas from a function already contained in an [`Arc`].
+    pub fn from_arc(draw: Arc<dyn Fn(&mut Scene, Size) + Send + Sync + 'static>) -> Self {
+        Self {
+            draw,
+            alt_text: None,
+        }
+    }
+
+    /// Set the text that will be used to communicate the meaning of the canvas to
+    /// those using screen readers.
+    ///
+    /// Users are strongly encouraged to set alt text for the canvas.
+    pub fn with_alt_text(mut self, alt_text: impl Into<String>) -> Self {
+        self.alt_text = Some(alt_text.into());
+        self
+    }
+}
+
+// --- MARK: WIDGETMUT ---
+impl Canvas {
+    /// Update the draw function
+    pub fn update_draw(
+        this: WidgetMut<'_, Self>,
+        draw: impl Fn(&mut Scene, Size) + Send + Sync + 'static,
+    ) {
+        Self::update_from_arc(this, Arc::new(draw));
+    }
+
+    /// Update the draw function
+    pub fn update_from_arc(
+        mut this: WidgetMut<'_, Self>,
+        draw: Arc<dyn Fn(&mut Scene, Size) + Send + Sync + 'static>,
+    ) {
+        this.widget.draw = draw;
+        this.ctx.request_render();
+    }
+}
+
+// --- MARK: IMPL WIDGET ---
+impl Widget for Canvas {
+    fn on_pointer_event(&mut self, _ctx: &mut EventCtx, _event: &PointerEvent) {}
+
+    fn accepts_pointer_interaction(&self) -> bool {
+        false
+    }
+
+    fn on_text_event(&mut self, _ctx: &mut EventCtx, _event: &TextEvent) {}
+
+    fn on_access_event(&mut self, _ctx: &mut EventCtx, _event: &AccessEvent) {}
+
+    fn register_children(&mut self, _ctx: &mut RegisterCtx) {}
+
+    fn update(&mut self, _ctx: &mut UpdateCtx, _event: &Update) {}
+
+    fn layout(&mut self, _ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
+        // use as much space as possible - caller can size it as necessary
+        bc.max()
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
+        (self.draw)(scene, ctx.size());
+    }
+
+    fn accessibility_role(&self) -> Role {
+        Role::Canvas
+    }
+
+    fn accessibility(&mut self, _ctx: &mut AccessCtx, node: &mut Node) {
+        // TODO: is this correct?
+        if let Some(text) = &self.alt_text {
+            node.set_description(text.clone());
+        } else {
+            node.clear_description();
+        }
```

Comment:
You can skip the `clear_description`. Node values are persisted between `accessibility` calls, you're rebuilding the node from scratch every time.

Reply:
I'm not sure what the first clause of the second sentence is trying to say. But I agree, you don't need clear_description, because by default there is no description.